sql: fix ALTER INDEX IF EXISTS for non-existing index

rafiss opened the pull request:
Previously, if ALTER INDEX IF EXISTS was used on an
unqualified index name that did not match any index, the database would
return an "index does not exist" error.

Now, the statement succeeds, but does nothing.

Also, test edge cases around ambiguous, non-existing, and/or unqualified
index names, in both alter_index and drop_index logic tests.

Fixes #42399 

Release note (bug fix): ALTER INDEX IF EXISTS no longer fails
when using an unqualified index name that does not match any existing
index. Now it is a no-op.

rafiss committed:
sql: fix ALTER INDEX IF EXISTS for non-existing index

Previously, if ALTER INDEX IF EXISTS (AIIE) was used on an
unqualified index name that did not match any index, the database would
return an "index does not exist" error.

Now, the statement succeeds, but does nothing.

Release note (bug fix): ALTER INDEX IF EXISTS no longer fails
when using an unqualified index name that does not match any existing
index. Now it is a no-op.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42797)
<!-- Reviewable:end -->

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> seems backport worthy!

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42797#-:-LudFBzt-ZxwKjsiZg43:b6ks6hg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @apantel)



<!-- Sent from Reviewable.io -->

apantel commented on pkg/sql/rename_index.go:
```diff
@@ -36,6 +36,10 @@ type renameIndexNode struct {
 func (p *planner) RenameIndex(ctx context.Context, n *tree.RenameIndex) (planNode, error) {
 	_, tableDesc, err := expandMutableIndexName(ctx, p, n.Index, true /* requireTable */)
```
Could you accomplish the same thing by passing requireTable=false here?

rafiss replied:
I actually ended up making this more similar to drop_index.go, and set `requireTable=!n.IfExists`

rafiss committed:
sql: add more tests to drop_index logic test

Test edge cases around ambiguous, non-existing, and/or unqualified
index names, to reach test parity with alter_index logic test.

Release note: None

apantel reviewed (commented):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42797#-:-LudoXmLEfJteEbrx84d:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @apantel and @jordanlewis)



<!-- Sent from Reviewable.io -->

rafiss commented:
thanks for the reviews!

bors r+

rafiss commented:
bors r-

rafiss commented:
bors r+
